### PR TITLE
GG-38005 Optimize topology history serialization

### PR DESCRIPTION
**Pull request by ibessonov**
https://ggsystems.atlassian.net/browse/GG-38005

### CODE REVIEW

File: modules/core/src/main/java/org/apache/ignite/configuration/TransactionConfiguration.java [hunk 1]
```diff
@@ -467,4 +468,44 @@ private static String[] transientSerializableFields(IgniteProductVersion ver) {
...
+        result = 31 * result + Objects.hashCode(dfltIsolation);
+        result = 31 * result + Objects.hashCode(dfltConcurrency);
+        result = 31 * result + Long.hashCode(dfltTxTimeout);
+        result = 31 * result + Long.hashCode(txTimeoutOnPartitionMapExchange);
+        result = 31 * result + Long.hashCode(deadlockTimeout);
+        result = 31 * result + pessimisticTxLogSize;
+        result = 31 * result + pessimisticTxLogLinger;
+        result = 31 * result + Objects.hashCode(tmLookupClsName);
+        result = 31 * result + Objects.hashCode(txManagerFactory);
+        result = 31 * result + Boolean.hashCode(useJtaSync);
```
[the remark this answers is not in the document]
Comment (ibessonov) [reply modules/core/src/main/java/org/apache/ignite/configuration/TransactionConfiguration.java:508]: I don't know, IDEA generated it for me

File: modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/internal/CompactedTopologyHistory.java [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+    private static final int DELTA_NODE = 1;
+
+    /** Indicated that node appeared in topology history before, and it's the same instance. */
+    private static final int SAME_NODE = 2;
+
+    /** Original topology history map. */
+    private SortedMap<Long, Collection<ClusterNode>> topHist;
+
+    /** The earliest known topology version. Cached in a field to save some time on a hot path. */
+    private long firstKey;
```
[the remark this answers is not in the document]
Comment (ibessonov) [reply modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/internal/CompactedTopologyHistory.java:65]: Sure

File: modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/internal/CompactedTopologyHistory.java [hunk 3]
```diff
@@ -0,0 +1,294 @@
...
+
+    /** Ctor. */
+    public CompactedTopologyHistory(@NotNull Map<Long, Collection<ClusterNode>> topHist) {
+        assert topHist instanceof SortedMap && !topHist.isEmpty() : topHist + " is invalid";
```
[the remark this answers is not in the document]
Comment (ibessonov) [reply modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/internal/CompactedTopologyHistory.java:69]: I would have to move the assertion to another place then, this parameter goes through several places that I don't want to modify

File: modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/internal/CompactedTopologyHistory.java [hunk 4]
```diff
@@ -0,0 +1,294 @@
...
+                Object consistentId = newNode.consistentId();
+                TcpDiscoveryNode oldNode = nodesMap.get(consistentId);
+
+                if (oldNode == null) {
+                    nodesMap.put(consistentId, newNode);
+
+                    out.write(NEW_NODE);
+                    writeTcpDiscoveryNode(out, newNode);
+                }
+                else if (oldNode != newNode) {
```
[the remark this answers is not in the document]
Comment (ibessonov) [reply modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/internal/CompactedTopologyHistory.java:114]: They might end up being equal, but in general they are different

File: modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/internal/CompactedTopologyHistory.java [hunk 5]
```diff
@@ -0,0 +1,294 @@
...
+    }
+
+    /**
+     * Returns a copy of {@code newNode}, but its attributes map contains a difference between attributes of
+     * {@code oldNode} and {@code newNode}.
+     */
+    private static TcpDiscoveryNode diff(TcpDiscoveryNode oldNode, TcpDiscoveryNode newNode) {
+        // It's important that we call "*.getAttributes()" instead of "*.attributes()", because the latter returns a
+        // wrapper that's a "view" on top of the original map, and it also filters out security attributes, meaning that
+        // it's literally missing some data.
```
[the remark this answers is not in the document]
Comment (ibessonov) [reply modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/internal/CompactedTopologyHistory.java:179]: Sure

File: modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/internal/CompactedTopologyHistory.java [hunk 6]
```diff
@@ -0,0 +1,294 @@
...
+        // Calculate updated keys.
+        // "forEach" here is not a stylistic choice, but rather an optimization. Attributes map that we iterate here is
+        // initialized with "Collections.unmodifiableMap", and its "*.entrySet()" iterator leads to a missive allocation
+        // of "Map.Entry" instances, which is noticeably slower than "forEach".
+        newAttrs.forEach((key, val) -> {
+            Object oldAttr = oldAttrs.get(key);
+
+            // We only call "containsKey" if "oldAttr" is "null". This way we avoid doing two lookups in the map because
+            // that's expensive.
+            boolean prevHasKey = oldAttr != null || oldAttrs.containsKey(key);
```
[the remark this answers is not in the document]
Comment (ibessonov) [reply modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/internal/CompactedTopologyHistory.java:194]: Old name from Ed, I'll change it

File: modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/internal/CompactedTopologyHistory.java [hunk 7]
```diff
@@ -0,0 +1,294 @@
...
+
+                        top.add(node);
+
+                        break;
+                    }
+
+                    case DELTA_NODE: {
+                        TcpDiscoveryNode newNode = readTcpDiscoveryNode(in);
+
+                        TcpDiscoveryNode oldNode = nodesMap.get(newNode.consistentId());
```
[the remark this answers is not in the document]
Comment (ibessonov) [reply modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/internal/CompactedTopologyHistory.java:261]: Why not

File: modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/internal/CompactedTopologyHistory.java [hunk 8]
```diff
@@ -0,0 +1,294 @@
...
+                        newNode.setAttributesNoCopy(newAttrs);
+
+                        nodesMap.put(newNode.consistentId(), newNode);
+
+                        top.add(newNode);
+
+                        break;
+                    }
+                    case SAME_NODE:
+                        top.add(nodesMap.get(in.readObject()));
```
[the remark this answers is not in the document]
Comment (ibessonov) [reply modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/internal/CompactedTopologyHistory.java:282]: Sure

File: modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/internal/TcpDiscoveryNode.java [hunk 9]
```diff
@@ -650,6 +660,35 @@ public TcpDiscoveryNode clientReconnectNode(Map<String, Object> nodeAttrs) {
             consistentId = consistentIdAttr != null ? consistentIdAttr : U.consistentId(addrs, discPort);
     }
 
+    /**
+     * For the exclusive use in {@link CompactedTopologyHistory}. Copies all non-transient fields into a new instance.
```
[the remark this answers is not in the document]
Comment (ibessonov) [reply modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/internal/TcpDiscoveryNode.java:664]: Yes

File: modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/internal/CompactedTopologyHistory.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (rpuch) [reply modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/internal/CompactedTopologyHistory.java:114]: Is it possible to use equals here then? If not, could you please explain the situation with a comment?

**Review by rpuch — APPROVED** [review]

**Review by rpuch — APPROVED** [review]

**Pull request merged**